client/core/trade.go

```diff
@@ -735,51 +737,15 @@ func (t *trackedTrade) isActive() bool {
 func (t *trackedTrade) matchIsRevoked(match *matchTracker) bool {
 	t.mtx.RLock()
 	defer t.mtx.RUnlock()
-	return match.MetaData.Proof.IsRevoked()
-}
-
-// Matches are inactive if: (1) status is complete, (2) it is refunded, or (3)
-// it is revoked and this side of the match requires no further action like
-// refund or auto-redeem. This should not be applied to cancel order matches.
-func (t *trackedTrade) matchIsActive(match *matchTracker) bool {
-	proof := &match.MetaData.Proof
-	// MatchComplete only means inactive if redeem request was accepted, but
-	// MatchComplete is set immediately after bcast for taker.
-	if match.MetaData.Status == order.MatchComplete && len(proof.Auth.RedeemSig) > 0 {
-		return false
-	}
-
-	// Refunded matches are inactive regardless of status.
-	if len(proof.RefundCoin) > 0 {
-		return false
-	}
-
-	// Revoked matches may need to be refunded or auto-redeemed first.
-	if proof.IsRevoked() {
-		// - NewlyMatched requires no further action from either side
-		// - MakerSwapCast requires no further action from the taker
-		// - (TakerSwapCast requires action on both sides)
-		// - MakerRedeemed requires no further action from the maker
-		// - MatchComplete requires no further action. This happens if taker
-		//   does not have server's ack of their redeem request (RedeemSig).
-		status, side := match.MetaData.Status, match.Match.Side
-		if status == order.NewlyMatched || status == order.MatchComplete ||
-			(status == order.MakerSwapCast && side == order.Taker) ||
-			(status == order.MakerRedeemed && side == order.Maker) {
-			t.dc.log.Tracef("Revoked match %v (%v) in status %v considered inactive.",
-				match.id, side, status)
-			return false
-		}
-	}
-	return true
+	return match.IsRevoked()
 }
 
 func (t *trackedTrade) activeMatches() []*matchTracker {
 	var actives []*matchTracker
 	t.mtx.RLock()
 	defer t.mtx.RUnlock()
 	for _, match := range t.matches {
-		if t.matchIsActive(match) {
```

Comment:
Although I think we'll end up accepting these changes, we need to reflect on the risk-reward of this diff because my view is that the changes are considerably more extensive and risky than necessary.

The problem to solve is fundamentally a DB issue:
- only taker
- match with final redeem broadcast
- not yet completed the `'redeem'` request
- restart of core does not reload the match **because it does not consider RedeemSig** in the DB

The severity of the issue is extremely low:
- very hard to hit (only taker, must shutdown after redeem bcast but before redeem request)
- very low impact (all txns bcast, no loss of funds, only the very lowest penalty)

Given the above, the way I consider the acceptable resolution is that we have very little tolerance for extensive changes or the possibility of new bugs, and the fix should be absolutely minimal.  I've commented on this line

`if t.matchIsActive(match) {` ==> `if !match.MetaData.Retired {`

because it is gets to what I think is too broad with the diff.  We are making changes to how `core` deals with order handling, both in "activeness" and revoke handling.  But afaict, nothing really needed to change in these places.  Now we have a new concern: keeping `Retired` consistent with all the conditions that were previously defined in `(*trackedTrade).matchIsActive`.

Considering this: you've added `Retired` to `dexdb.MatchMetaData`, so that solves the loading issue.  To support that we only needed to compute the correct `Retired` flag **when storing** Match data (a DB job), not necessarily change anything in core's operation.

I'll also note that yet another solution that was discussed that is even less involved, with no DB upgrade, is simply decoding match proof to check the redeem sig for taker matches with MatchSomplete status.  That solution is very very low risk, with only a minor (likely imperceptible) performance penalty when loading matches.  I ack'd that a `Retired` flag would probably be a better solution partly because of the inelegance of loading the match proof for all taker matches, hence this PR, but these were my concerns about it.

I won't ask to revert lots of code, but I would ask for comments regarding my conclusions above.  Were the changes in core strictly needed?  Are other problems solved beyond what I described, or are these more significant improvements to core than I have concluded?

Reply:
I do see that setting the `Retired` flag on db store only will resolve the issue at hand; didn't really think in that direction. One key reason for making the change wider was to have a more consistent definition of match activeness.

I'd earlier commented (much earlier) that having separate codes on db and core perform match activeness checks is prone to errors, as we've now seen and are trying to fix. While having the `Retired` flag work only with the db will solve the issue at hand, we'll still have 2 separate codes that determine match activeness, which could cause another issue of this sort to re-occur.

That said, I do also see that with the current approach taken in this PR, setting `match.Retired` is scattered all over the place. I had thought of making a single function that toggles that value, a function called when a match status progresses and when a match is retired; or any other scenario that changes a match activeness status. It could even still be part of the db match update method but that'd mean that simply changing a match status or retiring a match without an immediate successive call to db update will not set the Retired field appropriately. I'd initially gone the route of having a single code perform match activeness checks and setting the `Retired` field in one place but it requires the changes in #962 to go in first.

In summary, (apologies for that pretty long and hopefully not-confusing explanation), I agree with making this fix much simpler for the reasons you've mentioned above. I do still think we should have just one function that updates the `Retired` field appropriately and consistently and then both core and db can use that field to check for match activeness rather than having both pkgs maintain different codes for determining match activeness.